File: dcp-solutions/2024-09-22/solution.rb
```ruby
def longest_substring_with_k_distinct(s, k)
  if s.empty? || k == 0
    return 0
  end

  max_length = 0
  freq = Hash.new(0)
  left = 0

  for right in 0...s.size
    freq[s[right]] += 1
    while freq.size > k
      freq[s[left]] -= 1
      if freq[s[left]] == 0
        freq.delete(s[left])
      end
      left += 1
    end

    max_length = [max_length, right - left + 1].max
  end

  return max_length
end
```

File: dcp-solutions/2024-09-22/solution.go
```markdown
package main

import (
	"fmt"
)

func longestSubstring(s string, k int) int {
	if len(s) == 0 || k == 0 {
		return 0
	}

	var left int
	var charCount = make(map[rune]int)
	maxLength := 0

	for right, runeChar := range s {
		charCount[runeChar]++

		for size := len(charCount); size > k; {
			charCount[s[left]]--
			if charCount[s[left]] == 0 {
				delete(charCount, s[left])
			}
			left++
			size--
		}

		maxLength = max(maxLength, right-left+1)
	}

	return maxLength
}

func max(a int, b int) int {
	if a > b {
		return a
	} else {
		return b
	}
}

func main() {
	fmt.Println(longestSubstring("abcba", 2)) // prints: 5
}
```